test/e2e/storage/testsuites/volume_modify.go

```diff
@@ -225,11 +222,209 @@ func (v *volumeModifyTestSuite) DefineTests(driver storageframework.TestDriver,
+		}.AsSelector().String()
+		msg := "controller MODIFY_VOLUME capability is not reported"
+
+		err = e2eevents.WaitTimeoutForEvent(ctx, f.ClientSet, f.Namespace.Name, eventSelector, msg, f.Timeouts.ClaimProvision)
+		// Events are unreliable, don't depend on the event. It's used only to speed up the test.
+		if err != nil {
+			framework.Logf("Warning: did not get event about %s", events.ProvisioningFailed)
+		}
+
+		pvc, err := f.ClientSet.CoreV1().PersistentVolumeClaims(l.resource.Pvc.Name).Get(ctx, l.resource.Pvc.Name, metav1.GetOptions{})
+		framework.ExpectNoError(err, "could not re-read the pvc after event (or timeout)")
+		gomega.Expect(pvc.Status.Phase).To(gomega.Equal(v1.ClaimPending), "PVC phase isn't pending")
+	})
+
+	ginkgo.It("should not modify volume when csi driver does not support volume modification", func(ctx context.Context) {
+		if driver.GetDriverInfo().Capabilities[storageframework.CapModifyVolume] {
+			e2eskipper.Skipf("Driver %q supports VolumeAttributesClass - skipping negative test", driver.GetDriverInfo().Name)
+		}
+
+		l = local{}
+		l.config = driver.PrepareTest(ctx, f)
+		l.vac = MakeInvalidVAC(l.config)
+
+		ginkgo.By("Creating VolumeAttributesClass")
+		_, err := f.ClientSet.StorageV1beta1().VolumeAttributesClasses().Create(ctx, l.vac, metav1.CreateOptions{})
+		framework.ExpectNoError(err, "While creating VolumeAttributesClass")
+
+		ginkgo.By("Creating volume")
+		testVolumeSizeRange := v.GetTestSuiteInfo().SupportedSizeRange
+		l.resource = storageframework.CreateVolumeResource(ctx, driver, l.config, pattern, testVolumeSizeRange)
+		ginkgo.DeferCleanup(cleanup)
+
+		ginkgo.By("Creating a pod with dynamically provisioned volume")
+		podConfig := e2epod.Config{
+			NS:            f.Namespace.Name,
+			PVCs:          []*v1.PersistentVolumeClaim{l.resource.Pvc},
+			SeLinuxLabel:  e2epod.GetLinuxLabel(),
+			NodeSelection: l.config.ClientNodeSelection,
+			ImageID:       e2epod.GetDefaultTestImageID(),
+		}
+		pod, err := e2epod.CreateSecPodWithNodeSelection(ctx, f.ClientSet, &podConfig, f.Timeouts.PodStart)
+		ginkgo.DeferCleanup(e2epod.DeletePodWithWait, f.ClientSet, pod)
+		framework.ExpectNoError(err, "While creating test pod with VAC")
+
+		ginkgo.By("Attempting to modify PVC via VolumeAttributeClass")
+		newPVC := SetPVCVACName(ctx, l.resource.Pvc, l.vac.Name, f.ClientSet, setVACWaitPeriod)
+		l.resource.Pvc = newPVC
+		gomega.Expect(l.resource.Pvc).NotTo(gomega.BeNil())
+
+		ginkgo.By("Waiting for modification to fail by checking PVC events and status")
+		eventSelector := fields.Set{
+			"involvedObject.kind":      "PersistentVolumeClaim",
+			"involvedObject.name":      l.resource.Pvc.Name,
+			"involvedObject.namespace": l.resource.Pvc.Namespace,
+			"reason":                   events.VolumeModifyFailed,
+		}.AsSelector().String()
+		msg := "Unimplemented"
+
+		err = e2eevents.WaitTimeoutForEvent(ctx, f.ClientSet, f.Namespace.Name, eventSelector, msg, modifyVolumeWaitPeriod)
```

Comment:
What is the ideal total wait period for event and then modification failure? 

Current 20 Minutes total (2 x modifyVolumeWaitPeriod) is too long.

Reply:
We can remove all assertions on events if it can not speed up the test. Because events are unreliable, don't depend on the event. It's used only to speed up the test.

It may be a good practice to add event checks in unit tests to ensure that the events are emitted as expected.